gossip/src/duplicate_shred_handler.rs

```diff
@@ -90,17 +94,34 @@ impl DuplicateShredHandler {
     pub fn new(
         blockstore: Arc<Blockstore>,
         leader_schedule_cache: Arc<LeaderScheduleCache>,
+        bank_forks: Arc<RwLock<BankForks>>,
     ) -> Self {
         Self {
             chunk_map: HashMap::new(),
             validator_pending_proof_map: HashMap::new(),
-            last_root: 0,
+            cached_staked_nodes: None,
+            last_root: None,
+            last_root_cleaned: None,
+            cached_slots_in_epoch: 100,
             blockstore,
             leader_schedule_cache,
+            bank_forks,
             cleanup_count: CLEANUP_EVERY_N_LOOPS,
         }
     }
 
+    fn cache_root_info(&mut self) {
+        let last_root = self.blockstore.last_root();
+        if Some(last_root) != self.last_root {
+            self.last_root = Some(last_root);
+            if let Ok(bank_forks_result) = self.bank_forks.read() {
+                let root_bank = bank_forks_result.root_bank();
+                self.cached_staked_nodes = Some(root_bank.staked_nodes());
```

Comment:
since stakes only change per epoch, better to not take the `bank_forks` lock everytime the root changes.

Reply:
You are right, changed.